feat: introduce pInfo.GatingPlugin to filter out events more generally

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

We have a short cut path for the scheduling gate, which is not ideal because we're basically trying to build our scheduler generic; not depend on any in-tree plugins at all.

This PR introduces `pInfo.GatingPlugin`, which indicates the plugin name gating this Pod right now. Also, we create a pre-calculated map for which plugin is interested in which events. 
With them, we can implement an early `continue`, which is similar to the current short cut, but not specific to the scheduling gate plugin; This short cut path now benefits all the other PreEnqueue plugins too.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Follow up of #125527
We discussed we ideally wanted to make it general.

#### Special notes for your reviewer:

Here's the scheduler-perf result. 

**master**

```json
    {
      "data": {
        "Average": 1033.690384826998,
        "Perc50": 1042.0894467055684,
        "Perc90": 1296.9310473639357,
        "Perc95": 1577.0437629644223,
        "Perc99": 1577.0437629644223
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingWhileGated/1Node_10000GatedPods/namespace-5",
        "event": "not applicable",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 1068.7906572953373,
        "Perc50": 1076.0178436039005,
        "Perc90": 1375.026813022854,
        "Perc95": 1502.1168581823936,
        "Perc99": 1502.1168581823936
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingWhileGated/1Node_10000GatedPods_QueueingHintsEnabled/namespace-5",
        "event": "not applicable",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

**this branch**

```json
    {
      "data": {
        "Average": 1125.381883565056,
        "Perc50": 1063.7432906243237,
        "Perc90": 1238.9720202948656,
        "Perc95": 1760.1309819071496,
        "Perc99": 1760.1309819071496
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingWhileGated/1Node_10000GatedPods/namespace-5",
        "event": "not applicable",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
    {
      "data": {
        "Average": 1284.9195459589641,
        "Perc50": 1246.9339648710884,
        "Perc90": 1381.9668894552956,
        "Perc95": 1513.0076906180914,
        "Perc99": 1513.0076906180914
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingWhileGated/1Node_10000GatedPods_QueueingHintsEnabled/namespace-5",
        "event": "not applicable",
        "extension_point": "not applicable",
        "plugin": "not applicable",
        "result": "not applicable"
      }
    },
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

sanposhiho committed:
feat: introduce pInfo.GatingPlugin to filter out events more generally

sanposhiho commented:
/hold
To make sure this goes thru the approver's approve.

/cc @macsko

sanposhiho commented:
/cc @macsko @dom4ha @alculquicondor

macsko commented on pkg/scheduler/framework/types.go:
```diff
@@ -252,6 +261,8 @@ type QueuedPodInfo struct {
 	UnschedulablePlugins sets.Set[string]
 	// PendingPlugins records the plugin names that the Pod failed with Pending status.
 	PendingPlugins sets.Set[string]
+	// GatingPlugin records the plugin namesthat gated the Pod at PreEnqueue.
```
nit:
```suggestion
	// GatingPlugin records the plugin names that gated the Pod at PreEnqueue.
```

macsko commented:
/lgtm

dom4ha commented on pkg/scheduler/framework/types.go:
```diff
@@ -252,6 +261,8 @@ type QueuedPodInfo struct {
 	UnschedulablePlugins sets.Set[string]
 	// PendingPlugins records the plugin names that the Pod failed with Pending status.
 	PendingPlugins sets.Set[string]
+	// GatingPlugin records the plugin names that gated the Pod at PreEnqueue.
```
s/names/name

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -512,15 +527,18 @@ func (p *PriorityQueue) runPreEnqueuePlugins(ctx context.Context, pInfo *framewo
 		}
 		pInfo.UnschedulablePlugins.Insert(pl.Name())
+		pInfo.GatingPlugin = pl.Name()
+		pInfo.Gated = true
```
IMHO it was better when pInfo.Gated was set explicitly in one place. Note that the gatedBefore was captured prior changing the value.

Alternatively consider returning just the gatedPlugin name which would indicate whether the pod was gated or not

pInfo.GatingPlugin = runPreEnqueuePlugins(...)
pInfo.Gated = pInfo.GatingPlugin != ""

Actually, cannot pInfo.GatedPlugin replace pInfo.Gated? or make pInfo.Gated a function that returns pInfo.GatedPlugin != ""

sanposhiho replied:
> Actually, cannot pInfo.GatedPlugin replace pInfo.Gated?

We do have to know which plugin is gating this pod, so the answer is no.

> make pInfo.Gated a function that returns pInfo.GatedPlugin != ""

But, this is a good one.

ziyi-xie commented:
Hi, team. Sorry for the sudden message. I’m trying to gain a better understanding of scheduling_perf and what might be happening here and I would greatly appreciate your advice.
I’m running the following code on an Ubuntu 22.04 Linux VM with 4 CPUs, 8GB memory, and a 100GB hard disk, under various configurations.

> make test-integration WHAT=./test/integration/scheduler_perf ETCD_LOGLEVEL=warn KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingWhileGated/1Node_10000GatedPods"

I obtained the following results, which show the same trend you mentioned, but my tests actually failed. Could you let me know if this outcome is normal at your convenience please? 

> **QHint enabled**    master branch  vs.  this branch : 117 pods/s→ 103pods/s  performance decline
**QHint disabled**    master branch  vs.  this branch : 92 pods/s → 123 pods/s  performance improve

message generated by executing the code above
> make[1]: *** [Makefile:192: test] Error 1
> make[1]: Leaving directory '/home/xie/go/src/k8s.io/kubernetes'
> !!! [1018 05:41:12] Call tree:
> !!! [1018 05:41:12]  1: hack/make-rules/test-integration.sh:108 runTests(...)
> +++ [1018 05:41:12] Cleaning up etcd
> +++ [1018 05:41:13] Integration test cleanup complete
> make: *** [Makefile:215: test-integration] Error 1

sanposhiho commented:
@ziyi-xie Please open an issue and let's discuss there, @Gekko0114 said the same thing, so maybe it's a common issue.

sanposhiho commented:
@macsko @dom4ha (@ziyi-xie, FYI) 

I quickly checked scheduler-perf stuff, and it failed on my machine as well, although they correctly generated the results.

It's weird because if you see logs, it says both "FAIL" and "ok", also the exit code is 0.
I guess that's why we - maintainers - don't notice it, since the CI doesn't fail and we only look at the result files 😅 

```
ok  	k8s.io/kubernetes/test/integration/scheduler_perf	34.711s

=== Failed
=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling (unknown)
```

ziyi-xie commented:
Thank you so much for the quick response!
I’ve opened another issue as I'm not sure if this is the same one. If it turns out to be the same, I’m ready to close it anytime.
https://github.com/kubernetes/kubernetes/issues/128483

sanposhiho commented:
/hold

I don't touch this PR until https://github.com/kubernetes/kubernetes/pull/128170 is finalized
so that https://github.com/kubernetes/kubernetes/pull/128170 doesn't have to think about this upcoming change.

https://github.com/kubernetes/kubernetes/pull/128170 is the highest priority now, and shouldn't be interrupted by this chore change.

sanposhiho commented:
@macsko @dom4ha 
This PR is review for the review again. @macsko There's some changes after you put /lgtm so please recheck it.

macsko commented on pkg/scheduler/framework/types.go:
```diff
@@ -393,7 +407,6 @@ func (pqi *QueuedPodInfo) DeepCopy() *QueuedPodInfo {
 		InitialAttemptTimestamp: pqi.InitialAttemptTimestamp,
 		UnschedulablePlugins:    pqi.UnschedulablePlugins.Clone(),
 		PendingPlugins:          pqi.PendingPlugins.Clone(),
-		Gated:                   pqi.Gated,
```
Should't you copy the GatingPlugin instead?

sanposhiho commented on pkg/scheduler/backend/queue/scheduling_queue_test.go:
```diff
@@ -1676,22 +1679,27 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint(t *testing.
+			// In this case, the hintFn should be called as the pod is gated by foo plugin that is interested in the NodeAdd event.
+			hint: queueHintReturnQueue,
+			// and, as a result, this pod should be queued to activeQ.
 			expectedQ: activeQ,
```
Oops, corrected some mistakes in this test.

sanposhiho commented:
/retest

sanposhiho commented:
Well, I'm not sure if it's fair to aim to keep the current throughput because the current implementation is simple while only supports scheduling-gate vs this PR has to check more stuff while supports other plugins. Can we regard it as a fair trade-off?
I know the higher throughput it is, the better it is though, as we see, even though the throughput is dropped with this change, it's still higher than 300 pods/s, which we usually aim.

macsko commented:
> Do you have any alternative proposal that would allow us to get rid of the special casing for scheduling gates?

It looks like the map access has an impact on throughput:
https://github.com/kubernetes/kubernetes/blob/14fa32fe2771a618172bd6a4cc39c9352829c9df/pkg/scheduler/backend/queue/scheduling_queue.go#L1135

Couldn't we store the events list for GatingPlugin in pod info when setting GatingPlugin ([here](https://github.com/kubernetes/kubernetes/blob/14fa32fe2771a618172bd6a4cc39c9352829c9df/pkg/scheduler/backend/queue/scheduling_queue.go#L581))? Then, we'll get rid of that map access by accessing this events list directly.

sanposhiho commented on pkg/scheduler/framework/interface.go:
```diff
@@ -476,9 +476,12 @@ type QueueSortPlugin interface {
 // This is because such temporal errors cannot be resolved by specific cluster events,
```
resolving this comment because afterwards we decided to do a backoff in that case too.

sanposhiho commented:
Addressed the comments, PTAL again. Also, I retook the scheduler-perf result, and now the degradation is gone.

sanposhiho commented:
/retest

macsko commented on pkg/scheduler/framework/types.go:
```diff
@@ -380,20 +389,32 @@ type QueuedPodInfo struct {
 
 // DeepCopy returns a deep copy of the QueuedPodInfo object.
 func (pqi *QueuedPodInfo) DeepCopy() *QueuedPodInfo {
+	copyGatingPluginEvents := make([]ClusterEvent, len(pqi.GatingPluginEvents))
+	copy(copyGatingPluginEvents, pqi.GatingPluginEvents)
```
You can use `slices.Clone(pqi.GatingPluginEvents)` instead

sanposhiho replied:
Whoa, I didn't know it.

macsko commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -1099,22 +1133,9 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
-		// because PreEnqueue plugins apart from the scheduling gate plugin may change the gating status
-		// with these events.
-		if pInfo.Gated && pInfo.UnschedulablePlugins.Has(names.SchedulingGates) {
+		if pInfo.Gated() && (len(pInfo.GatingPluginEvents) == 0 || !event.MatchAny(pInfo.GatingPluginEvents)) {
```
`event.MatchAny(pInfo.GatingPluginEvents)` returns false for empty list of events, so it's not required to have `len(pInfo.GatingPluginEvents) == 0` here. But, it's up to you if we want an explicit check here.

sanposhiho replied:
OK, makes sense. Let's remove len() check here.

macsko commented:
/lgtm

dom4ha reviewed (commented):
> This way is a biiiit efficient, right? (only one write to `GatingPlugin`) 

I meant remove from `UnschedulablePlugins`. If we don't shrink the list, we could save from reconsidering those plugins in the future (assuming many plugins can block). If we have 4 plugins, each time one of the 4 unblocks, other 3 are reconsidered. I agree it's very theoretical, as we have only 3 possible plugins, so you can just leave a comment that it's not worth to remove it.

sanposhiho commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -116,11 +116,29 @@ func (pl *DefaultPreemption) PreEnqueue(ctx context.Context, p *v1.Pod) *framewo
 
+// isPodSchedulableAfterPodDeletion returns the queueing hint for the pod after the pod deletion event,
+// which always return Skip.
+// The default preemption plugin is a bit tricky;
```
Well, I'm not sure if I understood your idea right. But, that looks invalid.

This comment describes the following scenario, for example:
1. The Pod gets rejected by the node resource fit plugin and is moving to the preemption.
2. The Pod goes thru the async preemption process and is moving to the unsched Q.
3. At this point, the pod only has the node resource fit as the unsched plugins.
4. Some event happen, which the node resource fit's QHint return Queue, and the pod is tried to be moved to activeQ.
5. But, the async preemption isn't finished yet and the pod is blocked at PreEnqueue.
6. At this point, the pod gets the preemption plugin as the unsched plugins, in addition to the node resource fit.

And, after (6), we should move this pod when the node resource fit's QHint returns Queue; the preemption plugin's QHint only knows this pod can be schedulable with some Pod delete events, but it doesn't know (and doesn't have to know) which kind of Pod delete events are. That depends on which plugin(s) the Pod was originally rejected; in this example, it depends on the node resource fit plugin.
 
That's what this comment meant to say. So, 

> I'd however always return framework.Queue instead of framework.QueueSkip, as since there are other plugins blocking it, let others become the GatingPlugin.

It's not correct. The node resource fit shouldn't be GatingPlugin, and also I don't think it makes sense to always return Queue; that might make the Pod (that completed the async preemption) re-scheduled unnecessarily.

e.g., 

7. The async preemption is done and the pod is ungated.
8. Completely unrelated Pod is deleted (Pod/Delete event); the node resource fit returns QueueSkip, but the preemption plugin returns Queue. Also, the preemption plugin's PreEnqueue doesn't block the Pod anymore because the async preemption is finished.

By always returning QueueSkip from the preemption plugin, we can always let the node resource fit plugin decide whether this Pod should be retried or not.

dom4ha replied:
Makes sense. I originally assumed that other plugins also have PreEnqueue so that we should unblock as soon as possible and let other plugin become gated.

sanposhiho commented:
> I meant remove from UnschedulablePlugins. 

Right, that's a good point.

sanposhiho commented:
@macsko @dom4ha PTAL again when have time

macsko commented:
/lgtm

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -526,44 +540,63 @@ func (p *PriorityQueue) runPreEnqueuePlugins(ctx context.Context, pInfo *framewo
+		p.metricsRecorder.ObservePluginDurationAsync(preEnqueue, pl.Name(), s.Code().String(), p.clock.Since(startTime).Seconds())
+	}
+	if s.IsSuccess() {
+		return s
```
Hmmm, actually we probably shouldn't remove it from UnschedulablePlugins, because even if PreEnqueue succeeds, it doesn't mean the pod becomes schedulable. We still may need to watch for any cluster events in Qhint for this plugin.

dom4ha replied:
I struggle to see whether UnschedulablePlugins are only those which fail PreEnqueue, or all that made the pod unschedulable...

dom4ha commented:
/hold
/lgtm

sanposhiho commented:
@dom4ha 

> I think my suggestion was wrong (to remove from UnschedulablePlugins), because these are not only plugins for PreEnqueue, but for QHints of unschedulable pods in all phases. So even the PreEnqueue is passing, we still need to have QHints triggering as long as the pod is unschedulable.

The current implementation is not wrong, I think. Actually, I even think it's ok to remove all unschedulable plugins, once the pod is moved to backoffQ.

Let's say (1) the pod has `[PodAffinity, NodeAffinity]` in UnschedulablePlugins, (2) the pod gets `Queue` from NodeAffinity plugin, (3) the pod is moved to backoff, and (4) the pod is now rejected by the PreEnqueuePlugin-A.
In this case, the pod has to be get `Queue` from **PreEnqueuePlugin-A**, not the others (PodAffinity and NodeAffinity). Even if PodAffinity or NodeAffinity gives it `Queue`, it doesn't make the pod schedulable because the pod would be still blocked by PreEnqueuePlugin-A.

sanposhiho commented:
@dom4ha 

> I think my suggestion was wrong (to remove from UnschedulablePlugins), because these are not only plugins for PreEnqueue, but for QHints of unschedulable pods in all phases. So even the PreEnqueue is passing, we still need to have QHints triggering as long as the pod is unschedulable.

The current implementation is not wrong, I think. Actually, I even think it's ok to remove all unschedulable plugins, once the pod is moved to backoffQ.

Let's say (1) the pod has `[PodAffinity, NodeAffinity]` in UnschedulablePlugins, (2) the pod gets `Queue` from NodeAffinity plugin, (3) the pod is moved to backoff, and (4) the pod is now rejected by the PreEnqueuePlugin-A.
In this case, the pod has to be get `Queue` from **PreEnqueuePlugin-A**, not the others (PodAffinity and NodeAffinity). Even if PodAffinity or NodeAffinity gives it `Queue`, it doesn't make the pod schedulable because the pod would be still blocked by PreEnqueuePlugin-A. 

So, I think if a pod is rejected by PreEnqueue plugin, we probably should only put the PreEnqueue plugin in `UnschedulablePlugins`, removing all the others.
What do you think? cc @macsko

dom4ha commented:
> The current implementation is not wrong, I think. Actually, I even think it's ok to remove all unschedulable plugins, once the pod is moved to backoffQ.

When all PreEnqueue pass and we put a pod to active/backoff queues, then yes, it seems we can clean up unschedulable plugins list, since it will be rebuild in the next scheduling cycle from scratch (there is no other way to get back to the unschedulableQ).

What I meant though, the fact the PreEnqueue passed for a plugin X but it got gated on some other plugin Y, it doesn't meant we should no longer wait for QHints from the plugin X next time. It would be true assuming that new plugin Y has QHint that it would wake it up, but the example of async preemption in this PR is opposite - it has PreEnqueue, but it QHint always return QueueSkip. This is why I commented that it could always return Queue to check PreEnqueue each time.

> So, I think if a pod is rejected by PreEnqueue plugin, we probably should only put the PreEnqueue plugin in `UnschedulablePlugins`, removing all the others.
> What do you think? cc

There may be a set of a few PreEnqueue plugins which would still block a pod. So for instance blocking a pod on preemption PreEnqueue doesn't mean a pod should no longer run DRA PreEnqueue.

sanposhiho commented:
Okay, I was forgetting what I'm doing within the plugin in this PR.

So, I now have a contradiction;
1. I'm now saying we should make the queue only wait for GatingPlugin. i.e., clean up all plugins when the pod is being moved to BackoffQ (= after getting `Queue` from any of UnschedulablePlugins, and before running PreEnqueue)
2. On the other hand, for the async preemption case, we want to wait for UnschedulablePlugins (= plugins that _originally_ rejected the pod at the scheduling cycle) even after the preemption PreEnqueue plugin rejected the pod.

(If going with (1), (2) would die.)

But, I still don't believe always returning Queue (not Skip) in the preemption plugin is the right thing because [this reason](https://github.com/kubernetes/kubernetes/pull/127180#discussion_r1900356621) still applies.

Hmm, okay, for now, I'll just forget about what I was saying, and apply your first suggestion. (not to remove unschedplugins after PreEnqueue's execution)
Depending on how https://github.com/kubernetes/kubernetes/issues/130668 goes, _maybe_ we can change this async preemption's implementation not to use PreEnqueue, and then maybe we want to go with (1) afterwards. But, we can discuss later.

---

Answers for the other points:

> What I meant though, the fact the PreEnqueue passed for a plugin X but it got gated on some other plugin Y, it doesn't meant we should no longer wait for QHints from the plugin X next time.

It does. because the pod won't be schedulable (= un-gated) until plugin Y admits it.
Once plugin X passes this pod, plugin X is no longer related to this pod's schedulability at all. (again, the preemption plugin for the async preemption feature is the exception. they still want to wait for plugin-X, not plugin-Y)

> It would be true assuming that new plugin Y has QHint that it would wake it up
> the example of async preemption in this PR is opposite - it has PreEnqueue, but it QHint always return QueueSkip. This is why I commented that it could always return Queue to check PreEnqueue each time.

Yeah, so I _was_ thinking that PreEnqueue plugins should have QueueHint that returns Queue. (...but again, I know the preemption plugin has to return Skip at least for now)

dom4ha commented:
> So, for now, I'll just forget about what I was saying, and apply your first suggestion. (not to remove unschedplugins after PreEnqueue's execution) Depending on how #130668 goes, _maybe_ we can change this async preemption's implementation not to use PreEnqueue, and then maybe we want to go with (1) afterwards. But, we can discuss later.

Yes, this sounds wise to me considering the current state. Also it would be good to follow @macsko suggestion to clean it once we moved to backoff/activeQ.
 
> Answers for the other points:
> 
> > What I meant though, the fact the PreEnqueue passed for a plugin X but it got gated on some other plugin Y, it doesn't meant we should no longer wait for QHints from the plugin X next time.
> 
> It does. because the pod won't be schedulable (= un-gated) until plugin Y admits it. Once plugin X passes this pod, plugin X is no longer related to this pod's schedulability at all. (again, the preemption plugin for the async preemption feature is the exception. they still want to wait for plugin-X, not plugin-Y)

I could agree to that assuming that we had a way to enforce that each PreEnqueue implementation has the backing QHint that captures all events that could trigger reevaluation of PreEnqueue. Not sure how to construct such mechanism though, but I'd be cautious assuming all such cases will be captured during review (you already added a comment asking for that).

> Yeah, so I _was_ thinking that PreEnqueue plugins should have QueueHint that returns Queue. (...but again, I know the preemption plugin has to return Skip at least for now)

Sure, let's defer changing that once we go though the planned changes.

macsko commented:
I also agree that in the current state we should just keep the `UnschedulablePlugins` intact (or clear the whole list only when all PreEnqueue plugins pass, but it doesn't change anything - pod won't come back to the `unchedulablePods`).

> Once plugin X passes this pod, plugin X is no longer related to this pod's schedulability at all. (again, the preemption plugin for the async preemption feature is the exception. they still want to wait for plugin-X, not plugin-Y)

I feel like relying on cross-plugin dependencies is not a great approach long term. Probably, we could refine the async preemption implementation at some point, because the actual could lead to potential bugs like this one.

sanposhiho committed:
fix: not removing the plugin from the unsched plugins after PreEnqueue

sanposhiho commented:
Sorry for a long delay here, I rebased to resolve the conflict and addressed the point that we agreed (not to remove the plugin from unschedPlugins for QHint)

sanposhiho commented:
/retest

macsko commented:
/lgtm

Leaving the final stamp to @dom4ha

dom4ha commented:
/approve

sanposhiho commented:
/unhold

thanks
